### ISSUE TITLE
Support creation of minimal, no-solve application with Multiapps

### ISSUE THREAD

**Issue by permcody** [idaholab/moose#6126]
Per our conversation with @rppawlo:

We need the ability to create a minimal no-solve top level application with multiapps that can be appropriately controlled from an external application such as Vera.

**Comment by permcody** [comment]
@rppawlo 

After some twiddling, this appears to be the bare minimum input file for creating a "do nothing" main application:

```
[Mesh]
  type = GeneratedMesh
  dim = 2
  nx = 1
  ny = 1
[]

[Executioner]
  type = Transient
  num_steps = 20
  dt = 0.1
[]

[MultiApps]
  [./multi]
    type = TransientMultiApp
    app_type = MooseTestApp
    input_files = sub.i
    positions = '0 0 0'
  [../]
[]

[Problem]
  solve = false
[]
```

It appears that we really do need a Mesh currently in MOOSE. If I attempt to take the Mesh out, I get an error about FEProblem being NULL. Even though the mesh isn't used, it's required. We could look into taking it out but it might be more trouble than it's worth. I'll have to look into it and see...  It certainly won't consume much memory if you create a single element though.

The Executioner also currently needs to be there. Removing that will be rather difficult since it's assumed that after the application is setup, we can retrieve the one (and only one) executioner and call execute on it.  Of course, in general, we'd want an executioner to help with firing the remaining objects in the simulation. We could always create a custom Executioner for Vera. We don't need any variables, kernels or any other objects which is really nice. We can work with this.

**Comment by YaqiWang** [comment]
Just a thought, we could add a parameter in Problem, `non-solve` and let `Problem` be triggered and add the dummy mesh and executioner.

**Comment by permcody** [comment]
True, we can do all sorts of things to clean this up a little more. I'd be willing to do something like that if it helps the CASL folks out.

**Comment by friedmud** [comment]
I think that minimum file looks fine to me.  There are WAY too many things that depend on a Mesh existing.  It MUST be created.

If they want - they could add a new Mesh type to their application that automatically adds a single element to it and calls it a deal.  But that's more trouble than just providing this tiny Mesh block.

**Comment by rppawlo** [comment]
We are currently building the multiapp in-line in the code. There is no top level app input file at this point. We need direct access to call the transient multiapp solve time step (and resolve if needed). 

A few questions:

1. @permcody - you mentioned you would send info on building a fake top level app - something about a bison "assembly" if I recall correctly.  Is this still a viable option?
2. If restarting now goes through a top-level application, are there any issues with restarting but resetting some sub-apps.  I'm thinking in the context of solving multiple cycles of a nuclear reactor with bison fuel pins as sub-apps.  We want to reset some apps for fresh fuel and shuffle some apps around spatially.  By going through a top level app that was non-existent for us, I'm hoping I can still get at all the multiapp functionality and control.
3. I'd like to add a unit test to moose that has our use case of building a multiapp in-line instead of using an input file.  This would help us update for changes to the main driver.  How should I go about doing this?  Do you want a new application repo or can I just create a pull request that adds a new executable and test to the moose multiapp test directory?

**Comment by permcody** [comment]
On Mon, Jan 25, 2016 at 6:55 AM rppawlo <notifications@github.com> wrote:

> We are currently building the multiapp in-line in the code. There is no
> top level app input file at this point. We need direct access to call the
> transient multiapp solve time step (and resolve if needed).
>
> A few questions:
>
>    1. @permcody <https://github.com/permcody> - you mentioned you would
>    send info on building a fake top level app - something about a bison
>    "assembly" if I recall correctly. Is this still a viable option?
>
> Yes, I had mentioned that. I should clarify a bit though. The top level
application is not fake, it's a normal MooseApp. It's very much like the
one created in this input file. We turn of solving so even though there's
an app with a mesh and a system it's not being used. We looked into
disabling the mesh and it would be quite a bit of work so I would suggest
creating a single element mesh (either programmatically or just reading one
in). That mesh will get used to create a very small equation system so it's
a full fledged app, but then you don't need any active objects on it and
you don't need to solve it either. This is equivalent to the BISON "fuel
assembly" we used in our reactor simulation. We used to as a place holder
for the Multiapps underneath.


>
>    1. If restarting now goes through a top-level application, are there
>    any issues with restarting but resetting some sub-apps.
>
> There shouldn't be, this should work but I'm not certain we've used it for
awhile nor am I certain we have a test. I'll take a look but if you run
into any issues please let us know and we'll get them fixed.


>
>    1. I'm thinking in the context of solving multiple cycles of a nuclear
>    reactor with bison fuel pins as sub-apps. We want to reset some apps for
>    fresh fuel and shuffle some apps around spatially. By going through a top
>    level app that was non-existent for us, I'm hoping I can still get at all
>    the multiapp functionality and control.
>
> There are just too many things in MOOSE that rely on that top level
application existing, even if it doesn't do anything. I think this is the
way to go for the CASL simulation.


>
>    1. I'd like to add a unit test to moose that has our use case of
>    building a multiapp in-line instead of using an input file. This would help
>    us update for changes to the main driver. How should I go about doing this?
>    Do you want a new application repo or can I just create a pull request that
>    adds a new executable and test to the moose multiapp test directory?
>
> I think the best way would be to make an Example instead which is
basically a whole new application. We are currently testing for compilation
and a successfully exit code when running for the Examples but I plan to
expand that in the next week or two to run full unit tests on the Examples
looking at the solution file or other indicators. You're welcome to put in
a pull request along these lines.

Cody

**Comment by rppawlo** [comment]
OK.  I guess I need to create a top level application. I would like to do this inline.  In our original code, we would create a "throw away" top level app using a default parameter list to get access to  the factory to create the multiapp.  We then just used the multiapp directly (throwing away the original dummy app). Now it looks like we need to create a functioning top level app using the minimum input that @permcody specified earlier in this thread. So how do I create the inline equivalent parameter list to create the valid top level app for the input file you listed above? I guess I'm not sure how sublists for each section are created.  Do you have a simple example somewhere? The function I need to call is:
```
      Teuchos::RCP<MooseApp> app;
      {
        InputParameters app_params = AppFactory::instance().getValidParams("CASL_MOOSEApp");
        std::shared_ptr<CommandLine> command_line(new CommandLine(argc, argv));
        app_params.set<std::shared_ptr<CommandLine> >("_command_line") = command_line;
        app = Teuchos::rcp(AppFactory::instance().create("CASL_MOOSEApp","main",app_params,
              (*Teuchos::rcp_dynamic_cast<const Teuchos::MpiComm<int> >(bisonSubcomm)->getRawMpiComm())()));
      }
```
Th main issue is at runtime I generate a list of input files for the sub application so I don't have a static top level input file with the list of multiapp filenames for the MultiApp sublist (we use separate input files for each pin based on casl inputfile syntax).

**Comment by permcody** [comment]
We don't have a single top level constructor in MOOSE for setting up even that minimum skeleton file. The underlying assumption is that the Action System is going to fire off a bunch of Actions to build up the minimum system so we don't try to do everything just in one constructor. So normally you'd have to construct a mesh and set a few parameters in additional function calls. While that's doable, it's prone to breakage as we make adjustments over time. I'd entertain adding a slightly different set of parameters, or a different constructor for building the top level application to automatically do these extra steps for you. That's certainly not a problem for us.

However, it looks like you might be getting hung up elsewhere (the list of filenames in your second comment). I think we can handle that as well. There's no need to build up a dummy input file to create this top level application, it can all be done programmatically so you can continue to build your multiapps and add them to the system.

So to summarize, I'm thinking along these lines:

1.) Create the top level application like above but using a special parameter to tell MOOSE to go ahead and execute additional "default" or pre-canned actions to setup up a dummy mesh and set the parameters you need to make that application do nothing.

2.) Continue to build your Multiapps in your code programmatically and add them to the system like you are doing now.

3.) Control Execution

OK, so we need to implement 1) which should be straightforward. You are already doing 2). Finally, 3) may require that we add additional accessors or wrappers in MOOSE to help you control the backup, restore and stepping externally since the top level application will be lame. I'll probably need a little more information from you when you get there.

Am I on base here? Does this seem like a reasonable plan? If so, I can implement 1) for you.

**Comment by rppawlo** [comment]
That sounds great Cody. This is exactly what we need.  I can send you the current code we use to create the multiapp in an offline email so you see what we do and how we use the multiapp. For performing data transfers with DTK, we will still need access to the underlying multiapp object. What kind of time frame for implementing item 1?

**Comment by permcody** [comment]
I going on travel tomorrow but should be back late in the week. Should be
fairly easy to get 1) done this week or by early next week.

**Comment by rppawlo** [comment]
@permcody just wanted to check in on progress for item 1.

**Comment by permcody** [comment]
I haven't had a chance to work on it. I had a few fires to put out late
last week and am now enrolled in a NQA class for the duration of this week.
I'll try to get to this as soon as I can.
